raphtory/src/algorithms/algorithm_result.rs

```diff
@@ -0,0 +1,390 @@
+    ///
+    /// # Returns
+    ///
+    /// A sorted vector of tuples containing keys of type `H` and values of type `Y`.
+    pub fn sort_by_key(&self, reverse: bool) -> Vec<(H, Y)> {
+        let mut sorted: Vec<(H, Y)> = self.result.clone().into_iter().collect();
+        sorted.sort_by(|(a, _), (b, _)| {
+            if reverse {
+                b.cmp(a)
+            } else {
+                a.cmp(b)
+            }
+        });
+        sorted
+    }
+
+    /// Retrieves the top-k elements from the `AlgorithmResult` based on its values.
+    ///
+    /// # Arguments
+    ///
+    /// * `k`: The number of elements to retrieve.
+    /// * `percentage`: If `true`, the `k` parameter is treated as a percentage of total elements.
+    /// * `reverse`: If `true`, retrieves the elements in descending order; otherwise, in ascending order.
+    ///
+    /// # Returns
+    ///
+    /// An `Option` containing a vector of tuples with keys of type `H` and values of type `Y`.
+    /// If `percentage` is `true`, the returned vector contains the top `k` percentage of elements.
+    /// If `percentage` is `false`, the returned vector contains the top `k` elements.
+    /// Returns `None` if the result is empty or if `k` is 0.
+    pub fn top_k(&self, k: usize, percentage: bool, reverse: bool) -> Option<Vec<(H, Y)>> {
+        if percentage {
+            let total_count = self.result.len();
+            let k = (total_count as f64 * (k as f64 / 100.0)) as usize;
+            let sorted_result = self.sort_by_value(reverse);
+            Some(sorted_result.iter().cloned().take(k).collect())
+        } else {
+            let sorted_result = self.sort_by_value(reverse);
+            Some(sorted_result.iter().cloned().take(k).collect())
+        }
+    }
+
+}
+
+impl<H: Clone + Hash + Eq + Ord> AlgorithmResult<H, f64> {
+    /// Creates a new `AlgorithmResult` with floating-point values converted to `OrderedFloat`.
+    ///
+    /// # Arguments
+    ///
+    /// * `hashmap`: A `HashMap` with keys of type `H` and values of type `f64`.
+    ///
+    /// # Returns
+    ///
+    /// An `AlgorithmResult` with the `f64` values converted to `OrderedFloat<f64>`.
+    pub fn new_with_float(hashmap: HashMap<H, f64>) -> AlgorithmResult<H, OrderedFloat<f64>> {
+        let converted_hashmap: HashMap<H, OrderedFloat<f64>> = hashmap
+            .into_iter()
+            .map(|(key, value)| (key, OrderedFloat::from(value)))
+            .collect();
+        AlgorithmResult {
+            result: converted_hashmap,
+        }
+    }
+}
+
+impl<H, Y> AlgorithmResult<H, Y>
+    where
+        H: Clone + Hash + Eq + Ord,
+        Y: Clone + Ord + Hash + Eq
+{
+    /// Groups the `AlgorithmResult` by its values.
+    ///
+    /// # Returns
+    ///
+    /// A `HashMap` where keys are unique values from the `AlgorithmResult` and values are vectors
+    /// containing keys of type `H` that share the same value.
+    pub fn group_by(
```

Comment:
idk about whether group by can return an algo result
because then in the macro, the macro is returning itself
which it cannot do, because its building itself?